### ISSUE TITLE
Add context manager for disabling gradient checkpointing

### ISSUE THREAD

**Issue by timothybrooks** [facebookresearch/fairscale#771]
## 🚀 Feature
A method for disabling gradient checkpointing on-the-fly.

## Motivation

Training of some models requires use of `torch.autograd.grad()`. For example, this is commonly used for gradient penalty regularization when training GANs. Gradient checkpointing is not compatible with `torch.autograd.grad()` and only compatible with `torch.autograd.backward()`. It would be valuable to still allow gradient checkpointing for these networks, and provide a method to disabled checkpointing only when `grad()` is called.

## Pitch

```
from fairscale.nn.checkpoint.checkpoint_activations import checkpoint_wrapper, disable_checkpointing

...
module = checkpoint_wrapper(module)

# This line uses gradient checkpointing, as per usual.
y = module(x)

# However, this disables gradient checkpointing, and calls the original forward method.
with disable_checkpointing():
    y = module(x)
```

## Alternatives

Not use gradient checkpointing at all. This is unfavorable since in some cases `grad()` is used in a minor but necessary portion of training, and the other substantial calls to `backward()` are prohibited from using checkpointing. I have tried monkey patching this into fairscale but it gets pretty ugly and would be much cleaner to add directly to the library.

## Additional context

I believe this is easy to implement by adding an `is_checkpointing_disabled` flag to `thread_local`, which can be set inside a `disable_checkpointing()` context manager. The call to `_checkpointed_forward()` should then include:

```
if not torch.is_grad_enabled() or thread_local.is_checkpointing_disabled:
        return original_forward(module, *args, **kwargs)
```

If anyone sees an easier (temporary) workaround to this problem please let me know!

**Comment by min-xu-ai** [comment]
Thanks for the PR. I just merged it. It would be ideal to have a unit test that covers this but nobody has had time so far. I will file a new issue for that.
